app/services/aws/credentials_service.rb

```diff
@@ -8,46 +8,58 @@ def call(params = {})
+    TokenNotFoundError = Class.new(StandardError)
 
-    ROLE_SESSION_NAME = '3scale-porta'
+    DEFAULT_ROLE_SESSION_NAME = '3scale-porta'
+    IAM_KEYS = %i[access_key_id secret_access_key].freeze
+    STS_KEYS = %i[region role_arn role_session_name web_identity_token_file].freeze
 
     def initialize(params)
       @params = params
     end
 
-    # @return [Hash] Depending on the available connections, this method may return:
-    # When STS credentials are available: { credentials: Aws::STS::Types::AssumeRoleWithWebIdentityResponse };
-    # When STS credentials are not available: { web_identity_token_file: 'foo', role_name: 'bar' };
-    # When no configurations are available: An empy Hash{};
+    # Returns either IAM or STS credentials to be used on calls to S3
+    #
+    # @param [Hash] Hash containing AWS credentials related keys
+    # @return [Hash<Symbol, String>] with access_key_id and secret_access_key when IAM credentials are provided
+    # @return [Hash<Symbol, Aws::AssumeRoleWebIdentityCredentials>] when STS credentials are provided
+    # @raise [AuthenticationTypeError] if not enough params are provided
     def call
-      if sts_params.any? && sts_params.all? { |_key, value| value.present? } && web_identity_token_file_exists?
-        { credentials: sts_credentials }
-      else
+      if valid_credentials_with(iam_credentials, IAM_KEYS)
         iam_credentials
+      elsif valid_credentials_with(sts_credentials, STS_KEYS)
+        raise TokenNotFoundError, "web_identity_token_file was not found" unless web_identity_token_file_exists?
+
+        { credentials: Aws::AssumeRoleWebIdentityCredentials.new(sts_credentials) }
+      else
+        raise AuthenticationTypeError, "Either #{IAM_KEYS} or #{STS_KEYS} must be provided."
       end
     end
 
     private
 
-    def iam_credentials
-      params.slice(:access_key_id, :secret_access_key)
+    attr_reader :params
+
+    def valid_credentials_with(credentials, keys)
+      return false if credentials.empty?
+      return false if keys.difference(credentials.keys).any?
+
+      credentials.all? { |_key, value| value.present? }
     end
 
-    def sts_credentials
-      Aws::AssumeRoleWebIdentityCredentials.new(
-        web_identity_token_file: sts_params[:web_identity_token_file],
-        role_arn: sts_params[:role_name],
-        role_session_name: params[:role_session_name].presence || ROLE_SESSION_NAME
-      )
+    def web_identity_token_file_exists?
+      File.exist?(sts_credentials[:web_identity_token_file])
     end
 
-    def sts_params
-      params.slice(:web_identity_token_file, :role_name)
+    def iam_credentials
+      @iam_credentials ||= params.slice(*IAM_KEYS)
     end
 
-    def web_identity_token_file_exists?
-      File.exist?(sts_params[:web_identity_token_file])
+    def sts_credentials
+      @sts_credentials ||= params.slice(*STS_KEYS).tap do |credentials|
+        credentials[:role_session_name] ||= DEFAULT_ROLE_SESSION_NAME
```

Comment:
I think it makes sense to have a default path for `:web_identity_token_file` too. And then use `File.exists?` to validate that this file exists.

This will make it work configuration-less which is great. Setup cluster properly and skip adding any configuration values.

Reply:
I'm not sure on what could be a valid default value for it.
Operator will add a projected volume with the file, which means it'll be available for OpenShift deployments. A realistic value here would be `/var/run/secrets/openshift/serviceaccount/token`, which doesn't look to make sense with the application.
I'd say it's better not to assume it at all to avoid some level of coupling. The application simply receives the arguments and tries to understand which auth type it'll use, while guarding itself from crashes that could come from the external world (`File.exist?`).
What do you think?

Reply:
I think that we will have a documented way how to configure and there should be a way to figure out these settings  I think by cluster config or convention. Our app is deployed only in OpenShift so I don't see an issue in case that value is openshift specific.

If `File.exist?` crashes, then we will have a lot of bigger issues already. Also if there were no IAM credentials, then no STS credentials, then app will `raise` anyway. If `File.exists?` raises it will not make things much worse IMO.

For me ultimate goal is to make user experience seamless - need to do as little as possible.

For region - if we required it before, it can wait for another time. But this is a new functionality and is better to not introduce this later.

Reply:
> For me ultimate goal is to make user experience seamless - need to do as little as possible.

This is the point. Let me confirm with OPS what'll be the default path for the token :v:

Reply:
An update on this: @eguzki and I talked about whether or not the path to the token file will be configurable by our users. Since this will be an internal process, the operator will be responsible for creating the volume, putting the token file in the correct place and also provide the application with the correct location of the file, without the need of our users to set paths.
As this is related with OPS configuration, I'd avoid adding the default value to the application. Agreed?